RE: Add a button to upload an image to the primary hero image folder

xlisachan opened the pull request:
Fixes #14472 

A button to the primary hero admin has been added under the image selection section. Uploaded images will be stored in 'user-media/status/img/hero/featured'.

The test_admin.py test in /discovery/tests has also been updated.

Added section:
<img width="503" alt="Screen Shot 2020-06-03 at 8 37 14 AM" src="https://user-images.githubusercontent.com/28129806/83637331-73285080-a575-11ea-9f57-2ae7fb5df755.png">

After an image has been selected/uploaded from the computer:
<img width="507" alt="Screen Shot 2020-06-03 at 8 36 55 AM" src="https://user-images.githubusercontent.com/28129806/83637339-78859b00-a575-11ea-90a5-3022ba1c762f.png">

Please review when you have a moment. Thanks!

bobsilverberg commented:
Thanks for preparing the PR, @xlisachan! I have requested @eviljeff as a reviewer. In the future you can feel free to do this yourself when the PR is ready to be looked at or needs feedback.

xlisachan commented:
Will do, @bobsilverberg !

I have made updates to pass the CI checks. Please review when you have a moment, @eviljeff . Thanks!

eviljeff commented:
Hi @xlisachan a good start on getting something working! It looks like it would do what you wanted, but unfortunately this won't do what our Product team (Jorge, Scott, etc) have in mind.  They are looking for a way to add new photos to the pool of photos available, and then select from those photos in the shelf.

So, stepping back, I think what we need is a new django Model, called, say `PrimaryHeroImage`, and the new custom image property would be on _that_ model.  With a simple `class PrimaryHeroImageAdmin(admin.ModelAdmin)` to list and manage them.  (In hindsight I should have done this when I implemented the hero shelves a while back, but I thought it would only be a handful of images that would never change)

The next issue would then be to change the `image` property in `PrimaryHero` to be a foreign key to `PrimaryHeroImage`.  (There will be work to make it look good, to show images and thumbnails, but that's the basic).

xlisachan commented:
Thanks for the feedback, @eviljeff !  I'll make these updates and let you know if I have any questions!

bobsilverberg commented:
@xlisachan It sounds like we should change the description of this issue to match what's been suggested. Can you do that, please?

xlisachan commented:
@bobsilverberg, the description of this issue has been updated per your request.

xlisachan commented:
@eviljeff, the following updates have been made to provide admin the ability to upload images into the Primary Hero Image folder, and then select the image within the widget in Discovery Items.
- Removed `custom_image` property from `PrimaryHero` model
- Created `PrimaryHeroImage` model and added `custom_image` property to model
- Created `PrimaryHeroImageAdmin` to list and manage images
- Created `select_image` foreign key property to `PrimaryHero` model
- Added `def formfield_for_foreignkey` to admin
- Created `ImageChoiceField` to handle new image selection widget

I am still working on updating test_admin.py, as it has not passed the CI check. I will let you know if I need your guidance.

Please see the screenshots of the updated interface below:
- Location of 'Primary hero images' under the Discovery section where images for the primary hero are to be uploaded:
<img width="736" alt="Screen Shot 2020-06-12 at 11 55 56 AM" src="https://user-images.githubusercontent.com/28129806/84541123-bf396a80-acc4-11ea-968d-2dc2e8e5882e.png">

- 'Primary hero images' page:
<img width="996" alt="Screen Shot 2020-06-12 at 11 54 26 AM" src="https://user-images.githubusercontent.com/28129806/84541317-2a833c80-acc5-11ea-9be8-865d91e62fef.png">

- Widget within the Discovery Item where the primary hero image is selected:
<img width="1013" alt="Screen Shot 2020-06-12 at 11 54 11 AM" src="https://user-images.githubusercontent.com/28129806/84541238-0162ac00-acc5-11ea-92dc-a1d8a490a109.png">

Please review when you have a moment. Thank you!

eviljeff commented on src/olympia/hero/models.py:
```diff
@@ -88,7 +90,32 @@ def get_choices(self, *args, **kwargs):
+    preview_image.allow_tags = True
+
+
+class ImageChoiceField(forms.ModelChoiceField):
```
I think move this to admin.py

eviljeff commented on src/olympia/hero/models.py:
```diff
@@ -88,7 +90,32 @@ def get_choices(self, *args, **kwargs):
+
 class PrimaryHero(ModelBase):
+    select_image = models.ForeignKey(
+        PrimaryHeroImage, blank=True, null=True, on_delete=models.SET_NULL)
```
`blank` only really makes sense for text type fields -  it's not used for foreign keys

xlisachan replied:
I had forgotten to take out `blank` when I was trying to figure out how to not make `selected_image` a required field. Makes sense why that didn't work now, thanks for the clarification! It has now been removed. :)

eviljeff commented on src/olympia/hero/models.py:
```diff
@@ -88,7 +90,32 @@ def get_choices(self, *args, **kwargs):
+        return f'{self.custom_image}'
+
+    def preview_image(self):
+        return mark_safe('<img src="{}" width="150" />'.format(
```
You might need to copy the `Media` class declaration from `SecondaryHeroAdmin` to get the css to work.

eviljeff commented on src/olympia/hero/models.py:
```diff
@@ -88,7 +90,32 @@ def get_choices(self, *args, **kwargs):
+class ImageChoiceField(forms.ModelChoiceField):
+    def label_from_instance(self, obj):
+        return mark_safe(
+            '<img src="{}" width="150" height="120" />'.format(
```
same comment about defining a class in the html here - you may be able to share some css with the code above too.

eviljeff commented on src/olympia/hero/admin.py:
```diff
@@ -1,16 +1,47 @@
     model = PrimaryHero
-    fields = ('image', 'gradient_color', 'is_external', 'enabled')
+    fields = (
+        'select_image',
```
until we're ready to completely drop `image` we'll need to leave it in here (when we drop `image` the field will need removing from the model, and the serializer updating so `select_image` is served via the API)

xlisachan commented on src/olympia/hero/migrations/0007_auto_20200603_0207_squashed_0010_primaryhero_select_image.py:
```diff
@@ -0,0 +1,39 @@
+
+class Migration(migrations.Migration):
+
+    replaces = [('hero', '0007_auto_20200603_0207'), ('hero', '0008_remove_primaryhero_custom_image'), ('hero', '0009_primaryheroimage'), ('hero', '0010_primaryhero_select_image')]
```
Upon removing the `replaces` and `AlterField` lines, I had to update a migration file in discovery, which had a dependency on the 0009 hero migration file when I had added the 'Primary hero images' under the Discovery section in Admin. 

Local state of the database doesn't seem to be acting weird?? ... yet. I'll keep you posted.

eviljeff replied:
the weirdness might happen after you delete migrations 0008 to 0010 and then try to create a new migration.

xlisachan replied:
After squashing the migrations, I checked `showmigrations` which showed it was not utilizing 0007-0010 after the squash, so I removed those migration files. 

When I went ahead to make the migration to add the `image` property back into the fields in the PrimaryHero model, the migration had added the AlterField lines back in that were mentioned to be removed below, which I took out again before committing. No weirdness yet... but I'll let you know in case I just jinxed myself.

xlisachan commented on src/olympia/hero/admin.py:
```diff
@@ -1,16 +1,47 @@
+                'style': 'vertical-align: top'
+            })
+            kwargs['queryset'] = PrimaryHeroImage.objects
+            kwargs['empty_label'] = mark_safe("""
```
`def label_from_instance` did not seem to affect `empty_label`, so the inline html was needed to set the width.

eviljeff replied:
Can this be done with css instead?  Like, set the width for all the labels in css?

xlisachan replied:
I have updated the mark_safe to: 
`<div class="select-image-noimage">
    No image selected	                    
</div>`

xlisachan commented on src/olympia/hero/models.py:
```diff
@@ -88,7 +90,32 @@ def get_choices(self, *args, **kwargs):
 
+class PrimaryHeroImage(models.Model):
+    custom_image = models.ImageField(
+        upload_to='static/img/hero/featured',
```
I get the following error message with `upload_to=settings.HERO_FEATURED_IMAGE_PATH`:
`ImageField's 'upload_to' argument must be a relative path, not an absolute path.` Could I use `upload_to='code/storage/homepage/hero-featured-image'` instead?

eviljeff replied:
from looking at https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.FileField.upload_to it seems it expects a relative path to a folder under `MEDIA_ROOT` (`MEDIA_ROOT` in our case is `<STORAGE_ROOT>/shared_storage/uploads/`).

`upload_to`='hero-featured-image/' I suppose is good enough - it should end up in `code/storage/shared_storage/uploads/hero-featured-image/xxx.jpg` locally.

xlisachan commented:
Per your feedback, the following updates have been made:
- Migrations 0007-0010 have been squashed
- Unused migration files have been deleted
- Removed `replaces` and `AlterField` lines from squashed migration file
- Updated stylesheet and added Media class declaration in admin, which adds style to the 'No image selected' label, image labels, and images in PrimaryHeroImageAdmin
- Moved `ImageChoiceField` to admin.py
- Removed `blank` from `select_image` property
- Included `image` property back into model

As mentioned above, I received an error when I updating `upload_to` as the argument does not accept absolute paths. Please advise on whether `upload_to='code/storage/homepage/hero-featured-image'` can be used instead. Thanks!

xlisachan commented:
Per your feedback, the `upload_to` directory has been updated to 'hero-featured-image/'. Please review when you have a moment. Thanks!

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -290,7 +290,7 @@ def test_can_add_primary_hero_with_discovery_edit_permission(self):
                 'custom_addon_name': 'Xäxâxàxaxaxa !',
                 'recommendable': True,
                 'primaryhero-0-gradient_color': '#054096',
-                'primaryhero-0-image': 'Ubo@2x.jpg',
```
we still want `-image` to work, so this should stay

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -290,7 +290,7 @@ def test_can_add_primary_hero_with_discovery_edit_permission(self):
                 'recommendable': True,
                 'primaryhero-0-gradient_color': '#054096',
-                'primaryhero-0-image': 'Ubo@2x.jpg',
+                'primaryhero-0-select_image': 'Ubo@2x.jpg',
```
I'm surprised this actually works - is that what the form actually sends in the browser?  I would have thought the POST data (use devtools in browser) would be sending an int representing the `PrimaryHeroImage.id`

xlisachan replied:
You're right - it passes an int!

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -301,7 +301,7 @@ def test_can_add_primary_hero_with_discovery_edit_permission(self):
         assert item.custom_addon_name == 'Xäxâxàxaxaxa !'
         assert item.recommendable is True
         hero = PrimaryHero.objects.last()
-        assert hero.image == 'Ubo@2x.jpg'
```
same here, keep the `hero.image` assert

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -301,7 +301,7 @@ def test_can_add_primary_hero_with_discovery_edit_permission(self):
         assert item.recommendable is True
         hero = PrimaryHero.objects.last()
-        assert hero.image == 'Ubo@2x.jpg'
+        assert hero.select_image == 'Ubo@2x.jpg'
```
and assert here that `hero.select_image_pk == <the latest PrimaryHeroImage object>`.
Something like `assert hero.select_image_pk == image.pk` and _before_ the request is made you've done `image = PrimaryHeroImage.objects.create(custom_image='foo.jpg')` and then the `'primaryhero-0-select_image':` above is referencing the PrimaryHeroImage object.

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -301,7 +301,7 @@ def test_can_add_primary_hero_with_discovery_edit_permission(self):
-        assert hero.image == 'Ubo@2x.jpg'
+        assert hero.select_image == 'Ubo@2x.jpg'
         assert hero.gradient_color == '#054096'
         assert hero.disco_addon == item
```
We also need some tests that prove you can create, edit, and delete the `PrimaryHeroImage`s in django admin.  Basically, look at one of the existing TestXXX classes (e.g. `TestSecondaryHeroShelfAdmin`) and create a copy of it that tests `PrimaryHeroImage` instead.

xlisachan commented:
The following revisions were made in creating the tests to check that PrimaryHeroImage can be created, updated, and deleted in Django Admin:
- Updated `permissions.py`
- Added `custom_image` to `list_display` - (added `verbose_name='custom image path'`)
- Revised `test_admin.py` to include all tests

While all of the new tests passed, I wasn't 100% sure regarding the following lines:
- 562: I'm not sure if this is correct. The form data in DevTools did not have any information for the custom_image, as it returned for Discovery and SecondaryHeroShelves.
- 611/688: The other tests had `item.reload()`, but I kept receiving an error that it was not an attribute of PrimaryHeroImage.
- 649/655: I feel like there should be a sample filename for `custom_image`, but I kept getting an error if I included one

Please review when you have a moment, @eviljeff. Thanks!

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -546,6 +551,164 @@ def test_query_count(self):
+        response = self.client.post(
+            self.detail_url,
+            dict(self._get_heroform(str(item.id)), **{
+                'custom_image': 'föo.jpg',
```
this should be changing the `custom_image` value to verify the edit is working

xlisachan replied:
Oy, apologies for the oversight! I got caught up in making sure I included all the tests and checking they all passed *facepalm. I have updated the value and have it uploading a new file data instead of setting the filename.

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -546,6 +551,164 @@ def test_query_count(self):
+        response = self.client.post(
+            add_url,
+            dict(self._get_heroform(''), **{
+                'custom_image': '',
```
>  I feel like there should be a sample filename for custom_image, but I kept getting an error if I included one

What's the error?  Is it because you aren't actually including the file contents in the POST?  I suspect that may be necessary. 

I'm not sure what the request data structure would look like (if you select an image file in the form and then submit the form to upload it, devtools should show you) but it's probably something like `open(<path to image somewhere under tests />, 'rb')` and then `'upload': data` in the dict.

The `<path to image somewhere under tests/ >` you can probably use `olympia.amo.tests.test_helpers.get_image_path` for now.

xlisachan replied:
I think you were right about it being that I did not include the file contents. The error is gone now. :)

Could I use `olympia.amo.tests.test_helpers.get_uploaded_file`?

xlisachan commented:
Thanks for the kind feedback, @eviljeff! I have made the following updates:
- Updated the `custom_image` value in `def test_can_edit_with_discovery_edit_permission`
- Revised to upload file data as opposed to setting a filename
- Updated PrimaryHeroImage to inherit from `ModelBase`

As mentioned above, I used `olympia.amo.tests.test_helpers.get_uploaded_file` instead of `olympia.amo.tests.test_helpers.get_image_path`, please let me know if that works as well.

Regarding wanting the file to be deleted if the `PrimaryHeroImage` instance was deleted, in `def test_can_delete_with_discovery_edit_permission` the last line - `assert not PrimaryHeroImage.objects.filter(pk=item.pk).exists()` (line 628). Is checking that the file does not exist after the instance is deleted not the same?

Please review when you have a moment. Thanks!

eviljeff commented:
> As mentioned above, I used olympia.amo.tests.test_helpers.get_uploaded_file instead of olympia.amo.tests.test_helpers.get_image_path, please let me know if that works as well.

If it works, that's fine.

eviljeff commented on src/olympia/discovery/tests/test_admin.py:
```diff
@@ -546,6 +552,156 @@ def test_query_count(self):
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.count() == 1
+        item = PrimaryHeroImage.objects.get()
+        assert item.custom_image == 'hero-featured-image/transparent.png'
```
Can you assert that the file now exists in the uploads folder?  `assert os.path.exists(os.path.join(settings.MEDIA_ROOT, 'hero-featured-image', 'transparent.png'))` I guess.

xlisachan commented:
Per your feedback, I've asserted that the file now exists in the uploads folder into the test.
